src/installer/scripts.py

```diff
@@ -0,0 +1,120 @@
+
+_SCRIPT_TEMPLATE = """\
+# -*- coding: utf-8 -*-
+import re
+import sys
+from {module} import {import_name}
+if __name__ == "__main__":
+    sys.argv[0] = re.sub(r"(-script\\.pyw|\\.exe)?$", "", sys.argv[0])
+    sys.exit({func_path}())
+"""
+
+
+def _get_launcher_data(kind):
+    # type: (LauncherKind) -> Optional[builtins.binary_type]
+    if not kind:
+        return None
+    name = "{}.exe".format(kind)
+    return importlib_resources.read_binary(_scripts, name)
+
+
+def _is_executable_simple(executable):
+    # type: (builtins.binary_type) -> bool
+    if b" " in executable:
+        return False
+    shebang_length = len(executable) + 3  # Prefix #! and newline after.
+    # According to distlib, Darwin can handle up to 512 characters. But I want
+    # to avoid platform sniffing to make this as platform agnostic as possible.
+    # The "complex" script isn't that bad anyway.
+    return shebang_length <= 127
+
+
+def _quote_compat(s):  # pragma: no cover
+    # type: (builtins.text_type) -> builtins.text_type
+    """Crude implementation taken from shlex.quote().
+    """
+    return u"'" + s.replace(u"'", u"'\"'\"'") + u"'"
+
+
+def _build_shebang(executable, kind):
+    # type: (builtins.text_type, LauncherKind) -> builtins.binary_type
+    """Build a shebang line.
+
+    The non-launcher cases are taken directly from distlib's implementation,
+    which tries its best to account for command length, spaces in path, etc.
+
+    https://bitbucket.org/pypa/distlib/src/58cd5c6/distlib/scripts.py#lines-124
+    """
+    executable_bytes = builtins.ensure_binary(executable, encoding="utf-8")
+    if kind:  # The launcher can just use the command as-is.
+        return b"#!" + executable_bytes
+    if _is_executable_simple(executable_bytes):
+        return b"#!" + executable_bytes
+
+    # Shebang support for an executable with a space in it is under-specified
+    # and platform-dependent, so we use a clever hack to generate a script to
+    # run in ``/bin/sh`` that should work on all reasonably modern platforms.
+    # Read the following message to understand how the hack works:
+    # https://github.com/pradyunsg/installer/pull/4#issuecomment-623668717`
+
+    quote = getattr(shlex, "quote", _quote_compat)
+    quoted = builtins.ensure_binary(quote(executable), encoding="utf-8")
+    # I don't understand a lick what this is trying to do.
+    return b"#!/bin/sh\n'''exec' " + quoted + b' "$0" "$@"\n' + b"' '''"
+
+
+class Script(object):
+    __slots__ = ("name", "module", "attr")
+
+    def __init__(self, name, module, attr):
+        # type: (str, str, str) -> None
+        self.name = name
+        self.module = module
+        self.attr = attr
+
+    def __repr__(self):
+        # type: () -> str
+        return "Script(name={!r}, module={!r}, attr={!r}".format(
+            self.name, self.module, self.attr,
+        )
+
+    def build(self, executable, kind):
```

Comment:
Thinking about this a bit more, it feels wrong to include the gui/console dimension in ``kind``; the distinction is made in entry scripts sections, so the flag really should be set on the `Script` class instead. I want to make some modifications:

1. Make `Script.__init__()` accept an additional parameter to distinguish between a script in  `gui_scripts` and `console_scripts`.
2. Reduce the `kind` parameter here to only `""`, `"32"`, `"64"`, and `"arm"`. The launcher name is generated based on this kind parameter and the gui/console flag on the Script instance.

But @pradyunsg since you’ve already approved this PR, I’m fine to make the modification after this goes into master as well.

Reply:
Go ahead and make this change. I'm in no hurries to merge. :)